platform/posix/transport/src/plaintext_posix.c

```diff
@@ -147,21 +145,72 @@ int32_t Plaintext_Send( const NetworkContext_t * pNetworkContext,
                         const void * pBuffer,
                         size_t bytesToSend )
 {
-    int32_t bytesSent = 0;
+    int32_t bytesSent = -1, selectStatus = -1, getTimeoutStatus = -1;
+    struct timeval transportTimeout, selectTimeout;
+    socklen_t transportTimeoutLen;
+    fd_set writefds;
+
+    assert( pNetworkContext != NULL );
+    assert( pBuffer != NULL );
+
+    FD_ZERO( &writefds );
+    FD_SET( pNetworkContext->socketDescriptor, &writefds );
 
-    bytesSent = ( int32_t ) send( pNetworkContext->socketDescriptor, pBuffer, bytesToSend, 0 );
+    getTimeoutStatus = getsockopt( pNetworkContext->socketDescriptor,
+                                   SOL_SOCKET,
+                                   SO_SNDTIMEO,
+                                   &transportTimeout,
+                                   &transportTimeoutLen );
 
-    if( bytesSent < 0 )
+    if( getTimeoutStatus < 0 )
     {
-        logTransportError( errno );
+        /* Make #select return immediately if getting the timeout failed. */
+        selectTimeout.tv_sec = 0;
+        selectTimeout.tv_usec = 0;
+    }
+    else
+    {
+        selectTimeout.tv_sec = transportTimeout.tv_sec;
+        selectTimeout.tv_usec = transportTimeout.tv_usec;
+    }
+
+    /* Check if data can be written to the socket. */
+    selectStatus = select( pNetworkContext->socketDescriptor + 1,
+                           NULL,
+                           &writefds,
+                           &writefds,
+                           &selectTimeout );
```

Comment:
Just one thought. We currently have 2 timeouts. One set for the sockets and one for select. I think this can result in a worst case timeout of 2*timeout. I think it would be better to use only one. What are your thoughts?

Reply:
Back when we didn't use `select`, suppose the socket received data a millisecond before the timeout ended, then the data will likely not be received with the remaining time left, resulting in a timeout error.

In that sense, I think this new implementation is better. The timeout to receive data is no longer affected by how long the socket has been waiting on new data.